### PR TITLE
Disable spilling for distinct aggregation and pre-grouped aggregation

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Spilling for distinct aggregation has following problem:
The current implementation for distinct aggregation has assumption
that all the groups are kept in the hash table for the entire processing
so we can start yielding result immediately and output the new distinct
groups as new groups inserted in hash table. This assumption is no longer
true with spilling enabled. As we might turn on spilling in the middle of
input processing, there is no way to tell if a new inserted group is new 
distinct group or not as the hash table only holds a subset of groups. 
The proposed solution is described in [3263](https://github.com/facebookincubator/velox/issues/3263)

Spilling for pre-grouped keys has following problem:
The current pre-grouped key hash aggregation implementation starts to
yield aggregation result once after it processed all the input rows with
the same pre-grouped keys. It works like streaming aggregation and won't
hold much in-memory state. But in rare case that some particular
pre-grouped keys has very large number of rows, it can still trigger spilling.
Once it happens, we can not simply yield the result after process all the rows
with the same pre-grouped keys as the spilling is triggered in the middle of
row stream.
The proposed solution is described in [3264](https://github.com/facebookincubator/velox/issues/3264)

This PR disable the spilling for the above two hash aggregation use cases
and add unit tests to cover.

### CODE REVIEW

File: velox/core/QueryConfig.h [hunk 1]
```diff
@@ -94,11 +94,16 @@ class QueryConfig {
...
 
-  /// Preffered number of rows to be returned by operators from
+  /// Preferred number of rows to be returned by operators from
   /// Operator::getOutput.
   static constexpr const char* kPreferredOutputBatchSize =
       "preferred_output_batch_size";
 
+  /// Preferred number of  bytes to be returned by operators from
+  /// Operator::getOutput.
+  static constexpr const char* kPreferredOutputBatchByteSize =
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:104]: Would you extract this change into a separate PR and add it to documentation?

File: velox/core/QueryConfig.h [hunk 2]
```diff
@@ -94,11 +94,16 @@ class QueryConfig {
...
+      "preferred_output_batch_byte_size";
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:105]: nit: perhaps, preferred_output_batch_rows and preferred_output_batch_bytes ?

File: velox/exec/tests/AggregationTest.cpp [hunk 3]
```diff
@@ -1351,5 +1351,72 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
   }
 }
 
+TEST_F(AggregationTest, distinctWithSpilling) {
+  auto vectors = makeVectors(rowType_, 10, 100);
+  createDuckDbTable(vectors);
+  auto planNode =
+      PlanBuilder()
+          .values(vectors)
+          .aggregation(
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1360]: use singleAggregation method

File: velox/exec/tests/AggregationTest.cpp [hunk 4]
```diff
@@ -1351,5 +1351,72 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+              {"c0"}, {}, {}, core::AggregationNode::Step::kSingle, false)
+          .planNode();
+  auto tempDirectory = exec::test::TempDirectoryPath::create();
+  CursorParameters params;
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1364]: use AssertQueryBuilder

File: velox/exec/tests/AggregationTest.cpp [hunk 5]
```diff
@@ -1351,5 +1351,72 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+  for (auto& vector : vectors) {
+    auto* preGroupKeyFlatVector = vector->childAt(0)->asFlatVector<int64_t>();
+    for (int32_t row = 0; row < preGroupKeyFlatVector->size(); ++row) {
+      preGroupKeyFlatVector->set(row, val++ / 5);
+    }
+  }
+  createDuckDbTable(vectors);
+  auto planNode = PlanBuilder()
+                      .values(vectors)
+                      .orderBy({fmt::format("c0 ASC NULLS LAST")}, false)
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1390]: drop fmt::format

Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1390]: Why do we need orderBy if all keys are the same?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Thanks. Some comments below.

File: velox/core/QueryConfig.h (hunk 2 above)
Comment (xiaoxmeng) [reply velox/core/QueryConfig.h:105]: Better!

File: velox/core/QueryConfig.h (hunk 1 above)
Comment (xiaoxmeng) [reply velox/core/QueryConfig.h]: Will do per offline discussion.

File: velox/exec/tests/AggregationTest.cpp [hunk 6]
```diff
@@ -1351,5 +1351,59 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+                            .singleAggregation({"c0"}, {}, {})
+                            .planNode())
+                  .assertResults("SELECT distinct c0 FROM tmp");
+  auto stats = task->taskStats().pipelineStats;
+  // Verify that spilling is not triggered.
+  ASSERT_EQ(stats[0].operatorStats[1].spilledBytes, 0);
+}
+
+TEST_F(AggregationTest, preGroupedAggregationWithSpilling) {
+  auto vectors = makeVectors(rowType_, 10, 4);
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1374]: It looks like the generated vectors are not used as the loop below re-generates all the data. Consider using makeRowVector + makeFlatVector methods instead.

```
makeRowVector({
   // Pre-grouped key.
   makeFlatVector(1'000, [&] (auto /*row*/) { return val++ / 5; }),
   // Payload.
   makeFlatVector(1'000, [] (auto row) {return row}),
});
```

File: velox/exec/tests/AggregationTest.cpp [hunk 7]
```diff
@@ -1351,5 +1351,59 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+  // Make pre-grouping keys sorted.
+  int64_t val = 0;
+  for (auto& vector : vectors) {
+    auto* preGroupKeyFlatVector = vector->childAt(0)->asFlatVector<int64_t>();
+    for (int32_t row = 0; row < preGroupKeyFlatVector->size(); ++row) {
+      preGroupKeyFlatVector->set(row, val++ / 5);
+    }
+  }
+  createDuckDbTable(vectors);
+  auto tempDirectory = exec::test::TempDirectoryPath::create();
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1384]: nit: tempDirectory -> spillDirectory

File: velox/exec/tests/AggregationTest.cpp [hunk 8]
```diff
@@ -1351,5 +1351,59 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+                        {"c0"},
+                        {"sum(c2)"},
+                        {},
+                        core::AggregationNode::Step::kSingle,
+                        false)
+                    .planNode())
+          .assertResults("SELECT c0, c1, sum(c2) FROM tmp GROUP BY c0, c1");
+  auto stats = task->taskStats().pipelineStats;
+  // Verify that spilling is not triggered.
+  ASSERT_EQ(stats[0].operatorStats[1].spilledBytes, 0);
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1404]: [1] and [2] are a bit cryptic. Consider,

```
EXPECT_EQ(
      0,
      toPlanStats(task->taskStats())
          .at(aggNodeId)
          .spilledBytes);
```

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Looks good % small comments below.

**Comment by mbasmanova** [comment]
> Also add query property kPreferredOutputBatchByteSize to configure the operator output buffer.

@xiaoxmeng Would you update PR description to remove this sentence as this functionality has been moved into a different PR?

File: velox/exec/tests/AggregationTest.cpp [hunk 9]
```diff
@@ -1351,5 +1351,59 @@ TEST_F(AggregationTest, outputBatchSizeCheckWithSpill) {
...
+  auto tempDirectory = exec::test::TempDirectoryPath::create();
+  auto task =
+      AssertQueryBuilder(duckDbQueryRunner_)
+          .config(QueryConfig::kSpillEnabled, "true")
+          .config(QueryConfig::kAggregationSpillEnabled, "true")
+          .config(QueryConfig::kSpillPath, tempDirectory->path)
+          .config(QueryConfig::kTestingSpillPct, "100")
+          .plan(PlanBuilder()
+                    .values(vectors)
+                    .aggregation(
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/tests/AggregationTest.cpp:1393]: I need to pass pre-grouping keys so can't use singleAggregation in that case.
